File: 백준/Gold/2638. 치즈/치즈.py
import sys
import collections
input = sys.stdin.readline


def bfs(arr: list, n: int, m: int) -> int:
    visit = [[0 for j in range(m)] for i in range(n)]
    Q = collections.deque()
    del_set = set()


    visit[0][0] = 1
    Q.append([0,0])

    while Q:
        v = Q.popleft()

        for d in [[0,1],[1,0],[0,-1],[-1,0]]:
            ni = v[0] + d[0]
            nj = v[1] + d[1]
            if 0 <= ni < n and 0<= nj < m:
                if arr[ni][nj]: # 1 이면
                    if visit[ni][nj] < 0:
                        del_set.add((ni,nj))
                    visit[ni][nj] -= 1

                else:
                    if not visit[ni][nj]:
                        Q.append([ni,nj])
                        visit[ni][nj] = 1

    for i,j in del_set:
        arr[i][j] = 0

    if del_set:
        return True
    else:
        return False


if __name__ == '__main__':
    N, M = map(int,input().split())
    arr = [list(map(int,input().split())) for i in range(N)]
    res = 0
    
    while True:
        if bfs(arr, N, M):
            res += 1
            continue
        break
    
    print(res)

File: 백준/Gold/2636. 치즈/치즈.js
const bfs = (N, M, arr) =>{
  let visited = Array.from({length:N},()=>Array(M).fill(false));
  let melted = [];
  let Q = [[0,0]];
  visited[0][0] = true;
  let ni = 0, nj = 0;
  while (Q.length > 0){
    const [x, y] = Q.shift();
    for (const [n, m] of [[0,1],[1,0],[-1,0],[0,-1]]){
      ni = x+n;
      nj = y+m;
      if (0<=ni && ni < N && 0<=nj && nj < M && !visited[ni][nj]){
        if (arr[ni][nj] === 1){
          melted.push([ni, nj]);
          visited[ni][nj] = true;
        } else if (arr[ni][nj] === 0){
          Q.push([ni, nj]);
          visited[ni][nj] = true;
        }
      }
    }
  }
  let rtn = melted.length;
  for (const [i, j] of melted){
    arr[i][j] = 0;
  }
  return rtn;
}


const solve = (N, M, arr) => {
  let cheeze = arr.reduce((x, y)=>{
    return x + y.reduce((a, b)=> a+b, 0);
  }, 0)
  let rtn = 0;
  let cnt = 0;
  while (cheeze > 0){
    rtn = bfs(N, M, arr);
    cheeze -= rtn;
    cnt += 1
  }
  console.log(cnt);
  console.log(rtn);
}

const readline = require('readline');

const rl = readline.createInterface({
  input : process.stdin,
  output : process.stdout
})

let input = [];
rl.on('line', (line)=>{
  input.push(line);

  if (parseInt(input[0].split(' ')[0]) < input.length){
    rl.close();
  }
}).on('close', ()=>{
  const [N, M] = input[0].split(' ').map(Number);
  const arr = input.slice(1).map(e=>e.split(' ').map(Number));
  solve(N, M, arr);
})
